pgxn/neon/pagestore_smgr.c

```diff
@@ -1605,32 +1607,52 @@ neon_write(SMgrRelation reln, ForkNumber forknum, BlockNumber blocknum, const vo
 #endif
 {
 	XLogRecPtr	lsn;
+	RelKind relkind;
+	bool is_locked = false;
+	NRelFileInfo rinfo = InfoFromSMgrRel(reln);
 
 	switch (reln->smgr_relpersistence)
 	{
 		case 0:
-			/* This is a bit tricky. Check if the relation exists locally */
-			if (mdexists(reln, debug_compare_local ? INIT_FORKNUM : forknum))
+			relkind = get_cached_relkind(rinfo);
+			if (relkind == RELKIND_UNKNOWN)
+			{
+				/* We do not know relation persistence: let's determine it */
+				relkind = mdexists(reln, debug_compare_local ? INIT_FORKNUM : forknum) ? RELKIND_UNLOGGED : RELKIND_PERMANENT;
+				set_cached_relkind(rinfo, relkind);
```

Comment:
Is a race condition possible, where a concurrent backend changes the relkind from RELKIND_UNKNOWN to RELKIND_UNLOGGED_BUILD, between the `mdexists` and `set_cache_relkind` calls?

I guess not, because that would require that an unlogged build was started on a non-empty relation. But maybe add comments and a sanity check on that?